spec-tests/src/test/java/org/candlepin/spec/pools/RefreshPoolsSpecTest.java

```diff
@@ -847,6 +844,892 @@ public void shouldInvalidateEntitlementsWhenPoolQuantityIsReduced() {
+        assertEquals(1, ents.size());
+        assertEquals(1, ents.get(0).get("certificates").size());
+        JsonNode cert = ents.get(0).get("certificates").get(0);
+        assertNotNull(cert.get("serial"));
+        JsonNode serialBeforeUpdate = cert.get("serial").get("serial");
+
+        // Add content to product
+        ContentDTO content = adminClient.hosted().createContent(Content.random().vendor("vendor"));
+        prod = adminClient.hosted().addContentToProduct(prod.getId(), content.getId(), true);
+
+        refreshPools(adminClient, owner.getKey());
+        pools = adminClient.pools().listPoolsByOwner(owner.getId());
+        assertThat(pools).hasSize(1);
+
+        assertThat(adminClient.entitlements().getEntitlement(ents.get(0).get("id").asText()))
+            .isNotNull()
+            .extracting(EntitlementDTO::getCertificates)
+            .returns(1, Set::size)
+            .extracting(certs -> certs.iterator().next())
+            .extracting(CertificateDTO::getSerial)
+            .isNotNull()
+            .extracting(CertificateSerialDTO::getSerial)
+            .isNotEqualTo(serialBeforeUpdate.longValue());
+    }
+
+    @Test
+    @OnlyInHosted
+    public void shouldRegenerateEntsWhenDeletingContent() throws Exception {
+        ContentDTO content = adminClient.hosted().createContent(Content.random().vendor("vendor"));
+        ProductDTO prod = adminClient.hosted().createProduct(Products.random());
+        prod = adminClient.hosted().addContentToProduct(prod.getId(), content.getId(), true);
+        adminClient.hosted().createSubscription(Subscriptions.random(owner, prod));
+
+        refreshPools(adminClient, owner.getKey());
+        List<PoolDTO> pools = adminClient.pools().listPoolsByOwner(owner.getId());
+        assertThat(pools).hasSize(1);
+        PoolDTO pool = pools.get(0);
+
+        ConsumerDTO guestUser = createGuestUser();
+        ApiClient guestClient = ApiClients.ssl(guestUser);
+
+        JsonNode ents = guestClient.consumers().bindPool(guestUser.getUuid(), pool.getId(), 1);
+        assertEquals(1, ents.size());
+        assertEquals(1, ents.get(0).get("certificates").size());
+        JsonNode cert = ents.get(0).get("certificates").get(0);
+        assertNotNull(cert.get("serial"));
+        JsonNode serialBeforeUpdate = cert.get("serial").get("serial");
+
+        // remove content from product
+        adminClient.hosted().removeContentFromProduct(prod.getId(), List.of(content.getId()));
+
+        refreshPools(adminClient, owner.getKey());
+        pools = adminClient.pools().listPoolsByOwner(owner.getId());
+        assertThat(pools).hasSize(1);
+
+        assertThat(adminClient.entitlements().getEntitlement(ents.get(0).get("id").asText()))
+            .isNotNull()
+            .extracting(EntitlementDTO::getCertificates)
+            .returns(1, Set::size)
+            .extracting(certs -> certs.iterator().next())
```

Comment:
A more syntactically clear way of doing this would be using the assertion factory and `singleElement`:

```
assertThat(adminClient.entitlements().getEntitlement(ents.get(0).get("id").asText()))
    .isNotNull()
    .extracting(EntitlementDTO::getCertificates, as(collection(CertificateDTO.class)))
    .singleElement()
    .isNotNull()
    .extracting(CertificateDTO::getSerial)
    .isNotNull()
    .isNotEqualTo(serialBeforeUpdate.longValue());
```

We get the size check and the extraction in one line rather than two.

You'll need to add the following imports for the above code:
```
import static org.assertj.core.api.Assertions.as;
import static org.assertj.core.api.InstanceOfAssertFactories.collection;
```

This applies to all of these assertion chains throughout the suite.

Reply:
Perfect!! Thanks for this. I was really confused on why I could not use `.singleElement()` after extracting the set of CertificateDTOs and I ended up with that hacky approach. I'm still getting use to the assertj library, but I do like it a lot!

Reply:
:+1: